core/src/main/scala-3/pureconfig/generic/derivation/ProductConfigReaderDerivation.scala

```diff
@@ -0,0 +1,83 @@
+package pureconfig
+package generic
+package derivation
+
+import scala.compiletime.{constValue, constValueTuple, erasedValue, summonFrom, summonInline}
+import scala.compiletime.ops.int._
+import scala.deriving.Mirror
+import scala.util.chaining.*
+
+import pureconfig.error.{ConfigReaderFailures, ConvertFailure, KeyNotFound, UnknownKey, WrongSizeList}
+import pureconfig.generic.derivation.WidenType.widen
+
+trait ProductConfigReaderDerivation(fieldMapping: ConfigFieldMapping) { self: ConfigReaderDerivation =>
+  inline def derivedProduct[A](using m: Mirror.ProductOf[A]): ConfigReader[A] =
+    inline erasedValue[A] match {
+      case _: Tuple =>
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): ConfigReader.Result[A] =
+            for {
+              listCur <- asList(cur)
+              result <- readTuple[A & Tuple, 0](listCur.list, Nil)
+            } yield result
+
+          def asList(cur: ConfigCursor) =
+            cur.asListCursor.flatMap { listCur =>
+              if (constValue[Tuple.Size[A & Tuple]] == listCur.size)
+                Right(listCur)
+              else
+                listCur.failed(
+                  WrongSizeList(constValue[Tuple.Size[A & Tuple]], listCur.size)
+                )
+            }
+        }
+
+      case _ =>
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): ConfigReader.Result[A] =
+            for {
+              objCur <- cur.asObjectCursor
+              result <-
+                Labels
+                  .transformed[m.MirroredElemLabels](fieldMapping)
+                  .pipe(labels =>
+                    readTuple[m.MirroredElemTypes, 0](
+                      labels.map(objCur.atKeyOrUndefined(_)),
+                      labels.map(KeyNotFound.forKeys(_, objCur.keys))
+                    )
+                  )
+            } yield m.fromProduct(result)
+        }
+    }
+
+  inline def readTuple[T <: Tuple, N <: Int](
+      cursors: List[ConfigCursor],
```

Comment:
Any suggestions on how to do that? Because I am not sure I know how.

Reply:
Alright, so I got it working as long as I don't have any compile time verification of the cursor list size, and return an error of some kind if the list of cursors is too short (is there a correct error there already to use for that?).

Reply:
I realized only now that you're using this function both for both tuple-like and object-like configs, so you're right that we may not be able to get rid of this. I generally find this function heard to read because it assumes too many things about how it's called - reading its implementation I expected `cursors`, `keyNotFound` and the tuple type `T` to always have the same size, but in the tuple-like variant you call this with an empty `keyNotFound` list because you _know_ that cursors won't ever be undefined.

The function works though, and you already spent too much of your time on code style comments and nits from me. Let's keep this first version and think of improvements later 🙂

Reply:
Not able to get it working. I can make the list into an array, if that makes the N parameter anymore bearable for you?